Implement signing and signature verification.

BobWall23 opened the pull request:
Added GPG signature packet after the literal data packet in the encrypted data in ironsftp secure files. When decrypting, validate the signature.

BobWall23 committed:
Fix problem with read_pubkey_file error handling.

Read_pubkey_file was returning 0 even if it couldn't read the key file.
Fixed to return proper error and success values.
Also removed some crusty test / experimental code from iron-gpg.c

BobWall23 committed:
Fix phantom tab

BobWall23 committed:
Merge down from upstream

BobWall23 committed:
Big code munge to split iron-gpg.c

Split iron-gpg.c into several files in the iron/ subdir.
Detabbed all those files, converted "unsigned char" to "u_char".
Fixed Makefile.

BobWall23 committed:
Implement signing and signature verification.

Added GPG signature packet after the literal data packet in the
encrypted data in ironsftp secure files. When decrypting, validate the
signature.

zmre commented on iron/gpg-packet.c:
```diff
@@ -0,0 +1,1150 @@
+    int len = generate_gpg_tag_and_size(pubkey_pkt->tag, pubkey_pkt->len, buf);
+    SHA256_Update(&ctx, buf, len);
+    SHA256_Update(&ctx, sshbuf_ptr(pubkey_pkt->data), sshbuf_len(pubkey_pkt->data));
+
```
Does the second SHA256_Update not overwrite the &ctx? Does it append results? So this is the sha of the length concatenated with the sha of the data?

BobWall23 replied:
It's the SHA of (length concatenated with data).

zmre commented on iron/gpg-packet.c:
```diff
@@ -0,0 +1,1150 @@
+     * GPG hashes it, it expands the length out to four bytes instead of one. Grr.
+     */
+    if (uid_pkt != NULL) {
+        buf[0] = 0xb4;   //  uid_pkt->tag converted to a tag byte
```
0xb4 -- constant?

zmre commented on iron/gpg-packet.c:
```diff
@@ -0,0 +1,1150 @@
+    sshbuf_put_u8(body, GPG_SKALGO_AES256);
+    sshbuf_put_u8(body, GPG_SKALGO_AES192);
+    sshbuf_put_u8(body, GPG_SKALGO_AES128);
+    sshbuf_put_u8(body, 0);         //  No fourth preferred SK algorithm
```
So we put three different algorithm identifiers one after the next?

BobWall23 replied:
Yes - it's a fixed array of up to four preferred algorithms. I think there are similar arrays for PK algorithms and hash algorithms, and maybe for compression algorithms.

zmre commented on iron/gpg-packet.c:
```diff
@@ -0,0 +1,1150 @@
+	if (*(ptr++) != GPG_SIGCLASS_BINARY_DOC) return -3;
+	if (*(ptr++) != GPG_HASHALGO_SHA256) return -4;
+	if (*(ptr++) != GPG_PKALGO_RSA_ES) return -5;
+	memcpy(key_id, ptr, GPG_KEY_ID_LEN);
```
What are all these return codes? Maybe use constants?

BobWall23 replied:
We really want a file full of constants that no one is expected to do anything with? They are negative numbers that are different so I could tell where it was jumping out while troubleshooting.  I could just change them all to -1.

BobWall23 committed:
Fixed some more publicly visible funcs.

Moved iron_extension_offset into iron/util.c.
Prefixed additional public funcs with iron_.

BobWall23 committed:
Fixed bug with SEIPD packet length.

Added the length of the version # (one byte) twice, resulting in one
extra byte in the SEIPD packet length. GPG noticed and did not approve.

BobWall23 committed:
Fixed core dump fetching public RSA key.

Under certain conditions, when the RSA key was being fetched to encrypt
a message (public RSA key fetched to generate the passphrase to fetch
the secret key to sign the message), it would core dump dereferencing
the rsa pointer in the rsa key.

zmre commented on iron/gpg-key.c:
```diff
@@ -182,6 +182,36 @@ clamp_and_reverse_seckey(u_char * sk)
+ *  @param grip Place to write keygrip. At least SHA_DIGEST_LENGTH bytes.
+ */
+void
+generate_gpg_rsa_keygrip(const Key * rsa_key, u_char * grip)
```
test

BobWall23 replied:
Right back atcha.

BobWall23 committed:
Remove hard coded constant.

Generate header for UID packet instead of hard coding 0xb4 as the tag.
